# 失败的设计 - 受检异常
#Java 

Java 中的异常可以分为两大类别：受检异常与非受检异常。两者的区别在于：受检异常由编译器检查，遇到时抛给上层方法或捕获。而非受检异常在运行时发生，没有规定必须要捕获。

try-catch 异常处理是 Java 从 C++ 中借用过来的，但对于必须显式抛出或捕获受检异常是 Java 的补充。但是，这种概念在受 Java 影响的 Ruby、C#、Kotlin 语言中并没有得到体现，因此总的来说，这个概念是失败的。

受检异常存在着两个大问题：可伸缩性和版本控制。

## 版本控制

假设，我们定义了一个对外方法 foo，并且声明抛出异常 A、B、C：

```java
void foo() throws A, B, C;
```

发布方法后，我们在第二版的内容中想对 foo 方法功能进行调整，现在 foo 方法需要额外抛出一个异常 D。此时，如果在新版本中向 throws 子句中添加新异常会破坏已经存在的客户端代码。这种情况下有两种解决方案：

1. 创建一个全新的方法 foo2，新版本的功能在 foo2 方法中实现，此时 foo2 方法相比 foo 方法会额外抛出异常 D。
2. 在 foo 方法内部对异常 D 进行捕获，并将其转换为 A、B 或 C。

大多数情况下，客户端并不会太关心所调用的方法抛出何种异常。更多的情况是在调用处会有一个异常处理程序，该程序只会在异常抛出时通知问题发生并继续进行，之后通过 finally 代码块来保证程序会正常退出。因此，客户端实际上对处理异常并不感兴趣。

有时实现中抛出过多的受检异常，为了解决这些异常，人们会图简单做一些荒谬的事。例如，直接使用 `throws Exception` 来声明每个方法，这种声明对任何人都没有任何帮助。

> 人们认为异常的重要之处在于处理它们，但这并不是异常的重要内容。一个编写良好的应用程序中，try-finally 与 try-catch 的比率是 10:1。

## 可扩展性

在小型系统中，受检异常是很好的。例如，在读取文件时，捕获 FileNotFoundException 异常。但是在构建需要与多个不同子系统通信的大型系统时，如果每个子系统抛出 4-10 个异常，每当你聚合一个子系统时，你不得不声明更多可能需要捕获的异常。此时，你的异常处理处于一个失控的状态。

在大多情况下，受检异常处理会变得非常讨厌，以至于人们完全规避该功能。要么直接使用 `throws Exception` 语句，要么声明空的 catch 子句。在这些情况下，受检异常实际上大大降低了系统的质量。

